File: performable.php
<?php
  /**
   *  Usage
 
      $p = new Performable('XXXXXX'); // your account number
      $p->setId('123456');            // internal application id (optional)
      $p->setEmail('jane@doe.com');   // visitor's email address (optional)
      $p->setValue(124);              // purchase amount from shopping cart (optional)

      // additional properties you would like associated with the visitor
      $p->setCustomProperties(array('plan' => 'enterprise'));
      echo $p;

      // NOTE: requires json_encode function/library

   **/

  class Performable
  {

    var $accountId = "";
    var $identity = array();
    
    public function __construct($accountId)
    {
      $this->accountId = $accountId;
    }

    public function setEmail($email)
    {
      if(preg_match('|^\S+@\S+\.\S+$|i', $email)) {
        $this->identity['email'] = $email;
      }
    }

    public function setLastName($last_name)
    {
      if(!empty($last_name)) {
        $this->identity['last_name'] = $last_name;
      }
    }

    public function setFirstName($first_name)
    {
      if(!empty($first_name)) {
        $this->identity['first_name'] = $first_name;
      }
    }

    public function setFacebookId($facebookId)
    {
      if(!empty($facebookId)) {
        $this->identity['facebook_id'] = $facebookId;
      }
    }

    public function setFacebookUsername($facebook_username)
    {
      if(!empty($facebook_username)) {
        $this->identity['facebook'] = $facebook_username;
      }
    }

    public function setTwitterId($twitterId)
    {
      if(!empty($twitterId)) {
        $this->identity['twitter_id'] = $twitterId;
      }
    }

    public function setTwitterUsername($twitter_username)
    {
      if(!empty($twitter_username)) {
        $this->identity['twitter'] = $twitter_username;
      }
    }

    public function setId($id)
    {
      if(!empty($id)) {
        $this->identity['id'] = $id;
      }
    }

    public function setValue($value)
    {
      if(is_numeric($value)) {
        $this->identity['value'] = $value;
      }
    }

    public function setCustomProperties($arr) {
      if(!empty($arr)) {
        $this->identity = array_merge($this->identity, (array) $arr);
      }
    }

    public function __toString()
    {
      $lines = array();
      if(!empty($this->identity)) {
        $lines[] = '<script type="text/javascript">';
        $lines[] = 'var _paq = _paq || [];';
        $lines[] = sprintf('_paq.push(["identify", %s]);', json_encode($this->identity));
        $lines[] = '</script>';
      }
      $lines[] = sprintf('<script src="//d1nu2rn22elx8m.cloudfront.net/performable/pax/%s.js"></script>', $this->accountId);
      return join("\n", $lines);
    }
  }
?>

File: performable.py

# json is in the standard library as of python 2.6; fall back to
# simplejson if present for older versions.
try:
  import json
  assert hasattr(json, "dumps")
  _json_encode = lambda v: json.dumps(v)
except:
  try:
    import simplejson
    _json_encode = lambda v: simplejson.dumps(v)
  except ImportError:
    try:
      # For Google AppEngine
      from django.utils import simplejson
      _json_encode = lambda v: simplejson.dumps(v)
    except ImportError:
      raise NotImplementedError(
        "A JSON parser is required, e.g., simplejson at "
        "http://pypi.python.org/pypi/simplejson/")

import re

VALID_EMAIL = re.compile(r'^.+@.+\..+$')

class Performable(object):
  """
    Example Usage:
      x = Performable('XXXXXX')
      x.identify(id='123456', email='jane@performable.com')
      print x
  """

  def __init__(self, account):
    self.account = account
    self.identity = {}

  def identify(self, **kwargs):
    """ you should use one or more arguments such as:
      id - a stable internal user id
      email - user's email address
      facebook_id or facebook (for the username)
      twitter_id or twitter (for the username)

    e.g. x.identify(id='XYZ', twitter='eliast')
    """
    if 'email' in kwargs:
      if not VALID_EMAIL.match(kwargs.get('email')):
        del kwargs['email']

    self.identity.update(kwargs)

  def __str__(self):
    lines = []
    if self.identity:
      lines.append('<script type="text/javascript">')
      lines.append('var _paq = _paq || [];')
      lines.append('_paq.push(["identify", %s])' % _json_encode(self.identity))
      lines.append('</script>')
    lines.append('<script src="//d1nu2rn22elx8m.cloudfront.net/performable/pax/%s.js"></script>' % self.account)
    return '\n'.join(lines)


